### PR TITLE
T6944: adds option to enable switchdev mode on ethernet interface

### PR DESCRIPTION
**Pull request by nvollmar**
## Change Summary
Adds ethernet interface config option to enable switchdev mode.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
* https://vyos.dev/T6944

## Component(s) name
ethernet

## Proposed changes
Adds ethernet interface config option to enable switchdev mode.

## How to test
On a capable NIC set 
```
set interfaces ethernet eth0 switchdev
```
Get PCIe address of that NIC to verify state
```
ethtool -i eth0 | grep bus
devlink dev eswitch show pci/0000:01:00.1
```

## Smoketest result
Does not work on virtio_net adapters, requires manual testing on capable hardware

```
vyos@vyos:~$ /usr/libexec/vyos/tests/smoke/cli/test_interfaces_ethernet.py
test_add_multiple_ip_addresses (__main__.EthernetInterfaceTest.test_add_multiple_ip_addresses) ... ok
test_add_single_ip_address (__main__.EthernetInterfaceTest.test_add_single_ip_address) ... ok
test_add_to_invalid_vrf (__main__.EthernetInterfaceTest.test_add_to_invalid_vrf) ... ok
test_dhcp_client_options (__main__.EthernetInterfaceTest.test_dhcp_client_options) ... ok
test_dhcp_disable_interface (__main__.EthernetInterfaceTest.test_dhcp_disable_interface) ... ok
test_dhcp_vrf (__main__.EthernetInterfaceTest.test_dhcp_vrf) ... ok
test_dhcpv6_client_options (__main__.EthernetInterfaceTest.test_dhcpv6_client_options) ... ok
test_dhcpv6_vrf (__main__.EthernetInterfaceTest.test_dhcpv6_vrf) ... ok
test_dhcpv6pd_auto_sla_id (__main__.EthernetInterfaceTest.test_dhcpv6pd_auto_sla_id) ... ok
test_dhcpv6pd_manual_sla_id (__main__.EthernetInterfaceTest.test_dhcpv6pd_manual_sla_id) ... ok
test_eapol (__main__.EthernetInterfaceTest.test_eapol) ... ok
test_ethtool_evpn_uplink_tarcking (__main__.EthernetInterfaceTest.test_ethtool_evpn_uplink_tarcking) ... ok
test_ethtool_flow_control (__main__.EthernetInterfaceTest.test_ethtool_flow_control) ... ok
test_ethtool_ring_buffer (__main__.EthernetInterfaceTest.test_ethtool_ring_buffer) ... ok
test_interface_description (__main__.EthernetInterfaceTest.test_interface_description) ... ok
test_interface_disable (__main__.EthernetInterfaceTest.test_interface_disable) ... ok
test_interface_ip_options (__main__.EthernetInterfaceTest.test_interface_ip_options) ... ok
test_interface_ipv6_options (__main__.EthernetInterfaceTest.test_interface_ipv6_options) ... ok
test_interface_mtu (__main__.EthernetInterfaceTest.test_interface_mtu) ... ok
test_ipv6_link_local_address (__main__.EthernetInterfaceTest.test_ipv6_link_local_address) ... ok
test_move_interface_between_vrf_instances (__main__.EthernetInterfaceTest.test_move_interface_between_vrf_instances) ... ok
test_mtu_1200_no_ipv6_interface (__main__.EthernetInterfaceTest.test_mtu_1200_no_ipv6_interface) ... ok
test_non_existing_interface (__main__.EthernetInterfaceTest.test_non_existing_interface) ... ok
test_offloading_rfs (__main__.EthernetInterfaceTest.test_offloading_rfs) ... ok
test_offloading_rps (__main__.EthernetInterfaceTest.test_offloading_rps) ... ok
test_span_mirror (__main__.EthernetInterfaceTest.test_span_mirror) ... ok
test_speed_duplex_verify (__main__.EthernetInterfaceTest.test_speed_duplex_verify) ... ok
test_switchdev (__main__.EthernetInterfaceTest.test_switchdev) ... ok
test_vif_8021q_interfaces (__main__.EthernetInterfaceTest.test_vif_8021q_interfaces) ... ok
test_vif_8021q_lower_up_down (__main__.EthernetInterfaceTest.test_vif_8021q_lower_up_down) ... ok
test_vif_8021q_mtu_limits (__main__.EthernetInterfaceTest.test_vif_8021q_mtu_limits) ... ok
test_vif_8021q_qos_change (__main__.EthernetInterfaceTest.test_vif_8021q_qos_change) ... ok
test_vif_s_8021ad_vlan_interfaces (__main__.EthernetInterfaceTest.test_vif_s_8021ad_vlan_interfaces) ... ok
test_vif_s_protocol_change (__main__.EthernetInterfaceTest.test_vif_s_protocol_change) ... ok

----------------------------------------------------------------------
Ran 34 tests in 2148.235s
```

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation
- [X] I have updated the documentation accordingly

### CODE REVIEW

**Comment by nvollmar** [comment]
Tested on a system with Mellanox ConnectX-5 ✅

File: python/vyos/ifconfig/ethernet.py [hunk 1]
```diff
@@ -417,12 +418,38 @@ def set_ring_buffer(self, rx_tx, size):
             print(f'could not set "{rx_tx}" ring-buffer for {ifname}')
         return output
 
+    def set_switchdev(self, enable):
+        ifname = self.config['ifname']
+        if not bool(re.match(r"^eth[0-9]+$", ifname)):
+            print(f'switchdev mode is only possible on physical interfaces!')
+            return
+
+        addr, code = self._popen(f'ethtool -i {ifname} | grep bus-info | awk \'{{print $2}}\'')
```
Comment (sever-sever) [inline python/vyos/ifconfig/ethernet.py]: Will it be better to extend `Ethtool`  to get bus information?
https://github.com/vyos/vyos-1x/blob/current/python/vyos/ethtool.py
```
>>> from vyos.ethtool import Ethtool
>>> eth = Ethtool('eth0')
>>> eth.get_driver_name()
'virtio_net'
>>> 

```
Something like `get_bus_info` or `get_cpi_address`

Comment (nvollmar) [reply python/vyos/ifconfig/ethernet.py]: makes sense, will move it there

File: python/vyos/ifconfig/ethernet.py [hunk 2]
```diff
@@ -417,12 +418,38 @@ def set_ring_buffer(self, rx_tx, size):
...
+        if code != 0:
+            print(f'could not resolve PCIe address of {ifname}')
+            return
+
+        enabled = False
+        state, code = self._popen(f'devlink dev eswitch show pci/0000:01:00.1  | awk \'{{print $3}}\'')
```
Comment (sever-sever) [inline python/vyos/ifconfig/ethernet.py]: Why PCI is hardcoded?  Is it expected?

Comment (nvollmar) [reply python/vyos/ifconfig/ethernet.py]: this shouldn't have been in the commit, will fix that, should be `{addr}`

File: python/vyos/ifconfig/ethernet.py [hunk 3]
```diff
@@ -417,12 +418,38 @@ def set_ring_buffer(self, rx_tx, size):
...
+            print(f'could not resolve PCIe address of {ifname}')
+            return
+
+        enabled = False
+        state, code = self._popen(f'devlink dev eswitch show pci/{addr} | awk \'{{print $3}}\'')
+        if code == 0 and state == 'switchdev':
+            enabled = True
+
+        if enable and not enabled:
+            output, code = self._popen(f'/sbin/devlink dev eswitch set pci/{addr} mode switchdev')
```
Comment (c-po) [inline python/vyos/ifconfig/ethernet.py]: Why now with an absolute path and in line 433 no absolute path?

File: python/vyos/ifconfig/ethernet.py [hunk 4]
```diff
@@ -417,12 +418,38 @@ def set_ring_buffer(self, rx_tx, size):
             print(f'could not set "{rx_tx}" ring-buffer for {ifname}')
         return output
 
+    def set_switchdev(self, enable):
+        ifname = self.config['ifname']
+        if not bool(re.match(r"^eth[0-9]+$", ifname)):
```
Comment (c-po) [inline python/vyos/ifconfig/ethernet.py]: This seems to be superfluous, as handeld in the XML part

https://github.com/vyos/vyos-1x/blob/current/interface-definitions/interfaces_ethernet.xml.in#L17

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
I'd probably limit this feature to a NIC that was tested good, like the MLX CX5 and only allow it here.
Other NICs might join in the future. Do you know is the MLX CX6-LX is supported, too?

**Comment by nvollmar** [comment]
Works on CX4 and CX5, so don't see why it shouldn't work on CX6. AFAIK it also works on Intel cards like E810.

Could be done, but other offload methods also don't work on all cards and end in a warning if not supported.

File: python/vyos/ifconfig/ethernet.py (hunk 4 above)
Comment (nvollmar) [reply python/vyos/ifconfig/ethernet.py]: agree, removed

File: python/vyos/ifconfig/ethernet.py (hunk 3 above)
Comment (nvollmar) [reply python/vyos/ifconfig/ethernet.py]: fixed

File: python/vyos/ifconfig/ethernet.py [hunk 5]
```diff
@@ -417,6 +417,25 @@ def set_ring_buffer(self, rx_tx, size):
...
 
+    def set_switchdev(self, enable):
+        ifname = self.config['ifname']
+        addr, code = self._popen(f'ethtool -i {ifname} | grep bus-info | awk \'{{print $2}}\'')
+        if code != 0:
+            print(f'could not resolve PCIe address of {ifname}')
+            return
+
+        enabled = False
+        state, code = self._popen(f'/sbin/devlink dev eswitch show pci/0000:01:00.1  | awk \'{{print $3}}\'')
```
Comment (sever-sever) [inline python/vyos/ifconfig/ethernet.py]: PCI address still hardcoded

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: python/vyos/ifconfig/ethernet.py (hunk 5 above)
Comment (nvollmar) [reply python/vyos/ifconfig/ethernet.py]: sorry, fixed

**Review by dmbaturin — APPROVED** [review]
I'd be uneasy about hardcoded HCLs in scripts. We should certainly have lists of tested NICs in the docs. But if we don't allow people to try that option on new NICs, people will be unable to test them and report their findings.

We may add a list of knowingly problematic NICs, if any such NICs show up, and if trying to add the option makes the NIC entirely stop working until a hard reset or similar.

**Comment by nvollmar** [comment]
From my testing so far, more consumer NICs like I210-AT or virto just don't support it and fail to enable it (currently resulting in a message like enabling unsupported offload options). Afaik it must be supported by the NIC driver as well.

**Review by c-po — APPROVED** [review]

**Pull request merged**